cameo/strain_design/heuristic/evolutionary/optimization.py

```diff
@@ -806,3 +845,223 @@ def __init__(self, reaction_db=None, *args, **kwargs):
+            swapped = parent.clone(parent)
+            swapped.id = "{}_swap".format(swapped.id)
+            swapped.knock_out()
+            from_set = self.cofactor_swaps[from_to.index(True)]
+            to_set = self.cofactor_swaps[from_to.index(False)]
+            for index, metabolite in enumerate(to_set):
+                new_reactant = {metabolite: swapped.metabolites[from_set[index]]}
+                swapped.add_metabolites(new_reactant)
+            for metabolite in from_set:
+                metabolite._reaction.add(swapped)
+                swapped.pop(metabolite)
+            if swapped.check_mass_balance():
+                raise Exception('created reaction not mass balanced')
+            self.swapped_reactions[reaction_id] = swapped
+            if time_machine is not None:
+                time_machine(do=partial(self.add_reactions, [swapped]),
+                             undo=partial(self.remove_swap_reaction, reaction_id))
+            else:
+                self.add_reactions([swapped])
+
+    def remove_swap_reaction(self, reaction_id):
+        """
+        remove a swapped reaction and update the dictionary of reactions swaps
+
+        Parameters
+        ----------
+        reaction_id : string
+           identifier of the swapped reaction, must be member of `swapped_reactions`
+        """
+        self.remove_reactions([self.swapped_reactions[reaction_id]])
+        del self.swapped_reactions[reaction_id]
+
+    def swap_reaction(self, reaction_id, time_machine=None):
+        """
+        Knock-out the native reaction in and knock-in the reaction that has the co-factors swapped. Do it in reverse
+        if the native reaction is already knocker.
+
+        Parameters
+        ----------
+        reaction_id : string
+            the identifier of the reaction to swap
+        time_machine : TimeMachine
+            enable undo of the swap
+        """
+        from_reaction = self.reactions.get_by_id(reaction_id)
+        to_reaction = self.swapped_reactions[reaction_id]
+        if abs(from_reaction.lower_bound) < 1e-6 and abs(from_reaction.upper_bound) < 1e-6:
+            tmp_reaction = from_reaction
+            from_reaction = to_reaction
+            to_reaction = tmp_reaction
+
+        def do_change(reaction_a, reaction_b, ub, lb):
+            reaction_b.upper_bound = ub
+            reaction_b.lower_bound = lb
+            reaction_a.knock_out()
+
+        if time_machine is not None:
+            time_machine(do=partial(do_change, from_reaction, to_reaction,
+                                    from_reaction.upper_bound, from_reaction.lower_bound),
+                         undo=partial(do_change, to_reaction, from_reaction,
+                                      from_reaction.upper_bound, from_reaction.lower_bound))
+        else:
+            do_change(from_reaction, to_reaction, from_reaction.upper_bound, from_reaction.lower_bound)
+
+    def find_swappable_reactions(self, threshold=1e-6):
```

Comment:
I understood that. It makes sense with knockouts, because they have a cumulative effect. However, combinations of swaps might (say you change two reactions that are dependent via their cofactors) will work.